### ISSUE TITLE
Caught ClassNotFoundException caused by android:actionProviderClass?

### ISSUE THREAD

**Issue by MalcolmEvershed** [JakeWharton/ActionBarSherlock#540]
I ran the com.actionbarsherlock.sample.styled app on the Honeycomb MR2 emulator (API level 13) and I saw logcat output like:

<pre>
07-04 02:05:04.361: W/Resources(385): Converting to string: TypedValue{t=0x1/d=0x103011c a=-1 r=0x103011c}
07-04 02:05:04.371: W/MenuInflater(385): Ignoring attribute 'actionProviderClass'. Action view already specified.
07-04 02:05:04.494: W/Resources(385): Converting to string: TypedValue{t=0x1/d=0x103011c a=-1 r=0x103011c}
07-04 02:05:04.501: W/MenuInflater(385): Cannot instantiate class: @16974108
07-04 02:05:04.501: W/MenuInflater(385): java.lang.ClassNotFoundException: @16974108 in loader dalvik.system.PathClassLoader[/data/app/com.actionbarsherlock.sample.styled-2.apk]
07-04 02:05:04.501: W/MenuInflater(385):    at dalvik.system.PathClassLoader.findClass(PathClassLoader.java:251)
07-04 02:05:04.501: W/MenuInflater(385):    at java.lang.ClassLoader.loadClass(ClassLoader.java:540)
07-04 02:05:04.501: W/MenuInflater(385):    at java.lang.ClassLoader.loadClass(ClassLoader.java:500)
07-04 02:05:04.501: W/MenuInflater(385):    at com.actionbarsherlock.view.MenuInflater$MenuState.newInstance(MenuInflater.java:463)
07-04 02:05:04.501: W/MenuInflater(385):    at com.actionbarsherlock.view.MenuInflater$MenuState.readItem(MenuInflater.java:366)
07-04 02:05:04.501: W/MenuInflater(385):    at com.actionbarsherlock.view.MenuInflater.parseMenu(MenuInflater.java:147)
07-04 02:05:04.501: W/MenuInflater(385):    at com.actionbarsherlock.view.MenuInflater.inflate(MenuInflater.java:97)
07-04 02:05:04.501: W/MenuInflater(385):    at com.actionbarsherlock.sample.styled.MainActivity.onCreateOptionsMenu(MainActivity.java:88)
07-04 02:05:04.501: W/MenuInflater(385):    at android.support.v4.app._ActionBarSherlockTrojanHorse.onCreatePanelMenu(_ActionBarSherlockTrojanHorse.java:45)
07-04 02:05:04.501: W/MenuInflater(385):    at com.actionbarsherlock.ActionBarSherlock.callbackCreateOptionsMenu(ActionBarSherlock.java:556)
07-04 02:05:04.501: W/MenuInflater(385):    at com.actionbarsherlock.internal.ActionBarSherlockCompat.preparePanel(ActionBarSherlockCompat.java:483)
07-04 02:05:04.501: W/MenuInflater(385):    at com.actionbarsherlock.internal.ActionBarSherlockCompat.dispatchInvalidateOptionsMenu(ActionBarSherlockCompat.java:273)
07-04 02:05:04.501: W/MenuInflater(385):    at com.actionbarsherlock.internal.ActionBarSherlockCompat$1.run(ActionBarSherlockCompat.java:988)
07-04 02:05:04.501: W/MenuInflater(385):    at android.os.Handler.handleCallback(Handler.java:587)
07-04 02:05:04.501: W/MenuInflater(385):    at android.os.Handler.dispatchMessage(Handler.java:92)
07-04 02:05:04.501: W/MenuInflater(385):    at android.os.Looper.loop(Looper.java:132)
07-04 02:05:04.501: W/MenuInflater(385):    at android.app.ActivityThread.main(ActivityThread.java:4123)
07-04 02:05:04.501: W/MenuInflater(385):    at java.lang.reflect.Method.invokeNative(Native Method)
07-04 02:05:04.501: W/MenuInflater(385):    at java.lang.reflect.Method.invoke(Method.java:491)
07-04 02:05:04.501: W/MenuInflater(385):    at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:841)
07-04 02:05:04.501: W/MenuInflater(385):    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:599)
07-04 02:05:04.501: W/MenuInflater(385):    at dalvik.system.NativeStart.main(Native Method)
</pre>

This line of code in MenuInflater.java seems to cause the first "Converting to string" logcat output:

<pre>
itemActionProviderClassName = a.getString(R.styleable.SherlockMenuItem_android_actionProviderClass);
</pre>

Even though android:actionProviderClass isn't set on the menu item, itemActionProviderClassName is set to the string "@16974108". And of course there's no class with that name, so then the exception is thrown, etc.

My guess is that the issue is that ABS uses android:actionProviderClass which is only available in API Level 14 or newer. ... But then I don't know why this doesn't cause a problem on earlier platforms. Maybe the Android resource system behaves somewhat randomly when queried with android:foo attributes from the future.

Thanks.

**Comment by MalcolmEvershed** [comment]
My guess is that this same issue is responsible for #469.

**Comment by iNoles** [comment]
Actually, your problem is different because W/MenuInflater(385): Ignoring attribute 'actionProviderClass'. Action view already specified.

**Comment by MalcolmEvershed** [comment]
iNoles, when it says "Ignoring attribute 'actionProviderClass'", I don't think that implies that the problem is different because of the following:

The com.actionbarsherlock.sample.styled app does not specify actionProviderClass, so the error message is incorrect. Why is the error message incorrect? I broke into the debugger when that error message is emitted and the state was consistent with the problem described above (basically, itemActionProviderClassName was "@16974108", from being set a few lines above).

So basically, that error message is a red herring because actionProviderClass is not actually set in the menu resource. Thanks.

**Comment by MichaelRocks** [comment]
Attributes in `com.actionbarsherlock.R.attr.SherlockMenuItem` actually treated as some private attributes different from what they have to be. For instance, on Android 3.2 emulator `SherlockMenuItem_android_actionProviderClass` (`0x01010389`) is `android:attr/preferenceFrameLayoutStyle`. You can easily check this by calling `getResources().getResourceName(0x01010389)`.

The reason is that this attribute was added in 4.0, and in previous versions the value of the attribute was used by the internal attribute. Only public attributes preserve their values from version to version, because they are listed in the `public.xml` file. So declaring attributes with `android:` prefix is not a good idea.

**Comment by SimonVT** [comment]
It's not a bad idea since it ensures no changes to menu resources are necessary to make them work with ABS. Considering it doesn't crash, this is not really a major issue. To stop it from happening completely, simply checking the resource type should be enough (like we do with android:showAsAction).
